src/main/java/com/salesforce/functions/jvm/sdk/bulk/BulkApi.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright (c) 2021, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+package com.salesforce.functions.jvm.sdk.bulk;
+
+import com.salesforce.functions.jvm.sdk.bulk.builder.JobBuilder;
+import com.salesforce.functions.jvm.sdk.bulk.error.BulkApiException;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+
+/** Bulk API client to upload data in bulk to a Salesforce org. */
+public interface BulkApi {
+  /**
+   * Creates a job representing a bulk operation and its associated data that is sent to Salesforce
+   * for asynchronous processing. Provide job data via {@link #uploadJobData(String, File)}.
+   *
+   * @param job The job to create.
+   * @return The ID of the created job.
+   * @throws BulkApiException If an error occurred during job creation.
+   * @see #newJobBuilder(String, Operation)
+   */
+  @SuppressWarnings("unused")
+  String createJob(Job job) throws BulkApiException;
+
+  /**
+   * Retrieves detailed information about a job.
+   *
+   * @param jobId The ID of the job to get information about.
+   * @return Detailed information about a job.
+   * @throws BulkApiException If an error occurred while querying job info.
+   */
+  @SuppressWarnings("unused")
+  JobInfo getJobInfo(String jobId) throws BulkApiException;
+
+  /**
+   * Retrieves a list of successfully processed records for a completed job.
+   *
+   * <p>The returned file is a CSV file that all the records that encountered an error while being
+   * processed by the job. Each row corresponds to a successfully processed record.
+   *
+   * @see <a
+   *     href="https://developer.salesforce.com/docs/atlas.en-us.api_asynch.meta/api_asynch/get_job_successful_results.htm">Bulk
+   *     API 2.0 and Bulk API Developer Guide</a>
+   * @param jobId The ID of the job to get successful record results for.
+   * @return A CSV file with successful records.
+   * @throws BulkApiException If an error occurred while querying successful record results.
+   */
+  @SuppressWarnings("unused")
+  File getJobSuccessfulRecordResults(String jobId) throws BulkApiException;
```

Comment:
Should this return a `CsvFile` instead of a generic `File`?  Or will function users be have to parse the contents themselves?

Reply:
Yeah, it's the intention that they have to parse the resulting file themselves. It will be highly specific to them WRT to the contained data. The linked API docs don't even mention which kind of CSV to expect (delimiter, line-ending) so I assume it's the same as the input.

While certainly possible to implement complex CSV parsing logic, implementing good parsing in an abstract way over the specific customer CSV is complex endeavour. So I opted to go the most generic route.

Happy to be challenged on this though.

Reply:
Does `File` imply that our client library is saving this data to the filesystem? If so, that's a bit surprising to me. I would have guessed we'd use readable streams as the transport to the developer code. But maybe Java's File could work that way?

Reply:
Yes, and I agree that this seems weird. OTOH, the CSV that comes back from this call is potentially huge and might not easily fit in memory, especially in memory constrained systems like functions. By using the file-system, the SDK implementation can leverage streaming to write the data to disk in a streaming fashion without loading it all into memory first.

We could return an `InputStream` instead. The SDK implementation could then choose to still write to disk but expose the file via the stream or stream it directly from the TCP socket.

Reply:
Bulk API 2.0 supports several field delimiter characters: backquote (`), caret (^), comma, pipe (|), semicolon, and tab. The default delimiter is comma. Specify the delimiter to use when you create your job, using the columnDelimiter request field.

https://developer.salesforce.com/docs/atlas.en-us.api_asynch.meta/api_asynch/datafiles_prepare_csv.htm